substrate/primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -204,10 +291,212 @@ impl<V> OverlayedEntry<V> {
+				restore_append_to_parent(&mut parent.value, data, current_materialized);
+			}
+		}
+
+		if let Some(extrinsic) = at_extrinsic {
+			self.transaction_extrinsics_mut().insert(extrinsic);
+		}
+	}
+
+	/// Append content to a value, updating a prefixed compact encoded length.
+	///
+	/// This makes sure that the old version is not overwritten and can be properly
+	/// rolled back when required.
+	/// This avoid copying value from previous transaction.
+	fn append(&mut self, value: StorageValue, first_write_in_tx: bool, at_extrinsic: Option<u32>) {
+		if self.transactions.is_empty() {
+			self.transactions.push(InnerValue {
+				value: StorageEntry::Append {
+					data: AppendData::Data(value),
+					nb_append: 1,
+					materialized: None,
+					from_parent: false,
+				},
+				extrinsics: Default::default(),
+			});
+		} else if first_write_in_tx {
+			let parent = self.value_mut();
+			let (data, nb_append, materialized, from_parent) = match parent {
+				StorageEntry::None => (value, 1, None, false),
+				StorageEntry::Append { data, nb_append, materialized, from_parent: _ } => {
+					let AppendData::Data(data_buf) = data else {
+						unreachable!(
+							"append in last transaction and append in last transaction is data"
+						);
+					};
+					let mut data_buf = core::mem::take(data_buf);
+					*data = AppendData::MovedSize(data_buf.len());
+					StorageAppend::new(&mut data_buf).append_raw(value);
+					(data_buf, *nb_append + 1, *materialized, true)
+				},
+				StorageEntry::Some(prev) => {
+					// For compatibility: append if there is a encoded length, overwrite
+					// with value otherwhise.
+					if let Some(nb_append) = StorageAppend::new(prev).extract_nb_appends() {
+						// append on to of a simple storage should be avoided by any sane runtime,
+						// allowing a clone here.
+						// We clone existing data here, we could also change the existing value
+						// to an append variant to avoid this clone, but since this is should not
+						// happen in well written runtime (mixing set and append operation), the
+						// optimisation is not done here.
+						let mut data = prev.clone();
+						StorageAppend::new(&mut data).append_raw(value);
+						(data, nb_append + 1, Some(nb_append), false)
+					} else {
+						// overwrite, same as empty case.
+						(value, 1, None, false)
+					}
+				},
+			};
+			self.transactions.push(InnerValue {
+				value: StorageEntry::Append {
+					data: AppendData::Data(data),
+					nb_append,
+					materialized,
+					from_parent,
+				},
+				extrinsics: Default::default(),
+			});
+		} else {
+			// not first transaction write
+			let old_value = self.value_mut();
+			let replace = match old_value {
+				StorageEntry::None => Some((value, 1, None, false)),
+				StorageEntry::Some(data) => {
+					// Note that this code path is very unsafe (depending on the initial
+					// value if it start with a compact u32 we can have totally broken
```

Comment:
Indeed, the comment was assuming this data was set manually through a set_storage call in the first block (also true if from set_storage in a previous block), but yes when appending on on init it fetch a correctly encoding data and it is a normal code path.
Will try to reformulate that better.

Makes me realize the sequence is not really right here, we fetch value from previous block on first append. We should only fetch it when we access/materialize the value.
Changing behavior would need a new variant to AppendData containing init. But when use with this variant, all nb_append or MovedSize becomes relative to the to be fetch value.

Reply:
Realize the other day it is not a good idea in the context of the pr. Because it changes accesses to state, so it changes pov so it requires new host functions versionning and changes the spec. So probably better to keep this less efficient behavior (access to storage at first append call), lazy access would just have make things better in case we append and drop transaction with no reads.